### PR TITLE
chore: Update datafusion again

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
* keep up with dependencies
* get https://github.com/apache/arrow-datafusion/issues/5781

# Changes:
1. Update DataFusion pin to https://github.com/apache/arrow-datafusion/commit/bfffdbaba49a3f6772c4eff94befa95726de7195
2. Update for API changes

### CODE REVIEW

File: influxdb_iox/tests/query_tests/cases/in/date_bin.sql.expected [hunk 1]
```diff
@@ -109,7 +109,7 @@
 | physical_plan    | ProjectionExec: expr=[date_bin(Utf8("1 month"),cpu.time,Utf8("1970-12-31T00:15:00Z"))@0 as month, COUNT(cpu.user)@1 as COUNT(cpu.user)]    |
 |    |   AggregateExec: mode=FinalPartitioned, gby=[date_bin(Utf8("1 month"),cpu.time,Utf8("1970-12-31T00:15:00Z"))@0 as date_bin(Utf8("1 month"),cpu.time,Utf8("1970-12-31T00:15:00Z"))], aggr=[COUNT(cpu.user)]    |
 |    |     CoalesceBatchesExec: target_batch_size=8192    |
-|    |       RepartitionExec: partitioning=Hash([Column { name: "date_bin(Utf8(\"1 month\"),cpu.time,Utf8(\"1970-12-31T00:15:00Z\"))", index: 0 }], 4), input_partitions=4    |
+|    |       RepartitionExec: partitioning=Hash([date_bin(Utf8("1 month"),cpu.time,Utf8("1970-12-31T00:15:00Z"))@0], 4), input_partitions=4    |
```
Comment (alamb) [inline influxdb_iox/tests/query_tests/cases/in/date_bin.sql.expected:112]: This is due to https://github.com/apache/arrow-datafusion/pull/6708 -- which cleans up how expressions are displayed

File: influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected [hunk 2]
```diff
@@ -1387,6 +1387,7 @@ name: time_test
 +--------------------------+
 | tagKey                   |
 +--------------------------+
+|                          |
```
Comment (alamb) [inline influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: this does not look correct -- I need to figure out why this is diffing

File: iox_query/src/logical_optimizer/handle_gapfill/range_predicate.rs [hunk 3]
```diff
@@ -79,7 +79,7 @@ impl TreeNodeVisitor for TimeRangeVisitor {
 fn unwrap_alias(mut e: &Expr) -> &Expr {
     loop {
         match e {
-            Expr::Alias(inner, _) => e = inner.as_ref(),
+            Expr::Alias(Alias { expr, .. }) => e = expr.as_ref(),
```
Comment (alamb) [inline iox_query/src/logical_optimizer/handle_gapfill/range_predicate.rs:82]: due to https://github.com/apache/arrow-datafusion/pull/6795

File: influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected (hunk 2 above)
Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: Any hints about where to look @stuartcarnie ?

Comment (crepererum) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: Chiming in here: I would debug this using the following "hack":

1. trim the input file (`influxql_metadata.influxql` in this case) to a single failing query
2. run the query test w/ trace loging, e.g.: `TEST_INFLUXDB_IOX_CATALOG_DSN='postgres://postgres:mysecretpassword@localhost/iox_shared' TEST_INTEGRATION=1 RUST_BACKTRACE=1 CI=true RUST_LOG=trace cargo test -- query_tests::cases::influxql::influxql_metadata --exact`
3. you'll get the logical and physical plans logged at each optimization stage. store this somewhere
4. repeat step 3 w/o the re-pin / upgrade
5. compare outputs of step 3 and 4

From the top of my head I don't know what causes this. Looks like some differences in NULL handling. Maybe some unique/group-by now includes NULL keys?

Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: Thank you -- I will try and debug more today

Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: I followed your advice

Here is the expected output

```
-- Test Setup: InfluxQLSelectSupport
-- InfluxQL: SHOW TAG KEYS FROM time_test WHERE time >= '1990-01-01T00:00:00Z';
name: time_test
+--------------------------+
| tagKey                   |
+--------------------------+
| tt_tag                   |
| tt_tag_at_default_cutoff |
| tt_tag_late              |
+--------------------------+
````

Here is the actual output
```
-- Test Setup: InfluxQLSelectSupport
-- InfluxQL: SHOW TAG KEYS FROM time_test WHERE time >= '1990-01-01T00:00:00Z';
name: time_test
+--------------------------+
| tagKey                   |
+--------------------------+
|                          |
| tt_tag                   |
| tt_tag_at_default_cutoff |
| tt_tag_late              |
+--------------------------+
```
Plan before /after the upgrade look the same

```text
2023-07-07T16:12:54.409640Z DEBUG iox_query::exec::context: create_physical_plan: plan to run text=SortExec: expr=[iox::measurement@0 ASC NULLS LAST,tagKey@1 ASC NULLS LAST]
  ProjectionExec: expr=[time_test as iox::measurement, tagKey@0 as tagKey]
    CoalesceBatchesExec: target_batch_size=8192
      FilterExec: tagKey@0 IS NOT NULL
        UnnestExec
          ProjectionExec: expr=[make_array(CASE WHEN tt_tag@0 > 0 THEN tt_tag END, CASE WHEN tt_tag_at_default_cutoff@1 > 0 THEN tt_tag_at_default_cutoff END, CASE WHEN tt_tag_before_default_cutoff@2 > 0 THEN tt_tag_before_default_cutoff END, CASE WHEN tt_tag_late@3 > 0 THEN tt_tag_late END) as tagKey]
            AggregateExec: mode=Final, gby=[], aggr=[tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late]
              CoalescePartitionsExec
                AggregateExec: mode=Partial, gby=[], aggr=[tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late]
                  RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=2
                    ProjectionExec: expr=[tt_tag@1 as tt_tag, tt_tag_at_default_cutoff@2 as tt_tag_at_default_cutoff, tt_tag_before_default_cutoff@3 as tt_tag_before_default_cutoff, tt_tag_late@4 as tt_tag_late]
                      CoalesceBatchesExec: target_batch_size=8192
                        FilterExec: time@0 >= 631152000000000000
                          ParquetExec: file_groups={2 groups: [[1/1/4710107baa934df3ee81038c93840af52a74c16e2c4203154b678aec3d59429f/bd13a628-c347-49f3-9633-0db6ec1a149e.parquet], [1/1/5449d2517aad579ad5a02ebe782d6a4e9e9d5796feb3a1103987b40ba61cf7a2/7ce75c62-19ea-4087-9dec-1b2e470f7130.parquet]]}, projection=[time, tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late], predicate=time@0 >= 631152000000000000, pruning_predicate=time_max@0 >= 631152000000000000
```

Plan After the upgrade
```
2023-07-07T16:08:34.882455Z DEBUG iox_query::exec::context: create_physical_plan: plan to run text=SortExec: expr=[iox::measurement@0 ASC NULLS LAST,tagKey@1 ASC NULLS LAST]
  ProjectionExec: expr=[time_test as iox::measurement, tagKey@0 as tagKey]
    CoalesceBatchesExec: target_batch_size=8192
      FilterExec: tagKey@0 IS NOT NULL
        UnnestExec
          ProjectionExec: expr=[make_array(CASE WHEN tt_tag@0 > 0 THEN tt_tag END, CASE WHEN tt_tag_at_default_cutoff@1 > 0 THEN tt_tag_at_default_cutoff END, CASE WHEN tt_tag_before_default_cutoff@2 > 0 THEN tt_tag_before_default_cutoff END, CASE WHEN tt_tag_late@3 > 0 THEN tt_tag_late END) as tagKey]
            AggregateExec: mode=Final, gby=[], aggr=[tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late]
              CoalescePartitionsExec
                AggregateExec: mode=Partial, gby=[], aggr=[tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late]
                  RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=2
                    ProjectionExec: expr=[tt_tag@1 as tt_tag, tt_tag_at_default_cutoff@2 as tt_tag_at_default_cutoff, tt_tag_before_default_cutoff@3 as tt_tag_before_default_cutoff, tt_tag_late@4 as tt_tag_late]
                      CoalesceBatchesExec: target_batch_size=8192
                        FilterExec: time@0 >= 631152000000000000
                          ParquetExec: file_groups={2 groups: [[1/4/5c7be769220c44c2a96f699c1484750595edc1e07de516d5132b9796339353a8/c1788694-a185-4acd-b26c-2d7959cc6c9d.parquet], [1/4/972dcd6728d71c567b7f705c5e22ea9f1dc7aed8d3fda0042885fc5725143509/1b23b8d0-8e8e-4f3c-abca-9ab94060d2dd.parquet]]}, projection=[time, tt_tag, tt_tag_at_default_cutoff, tt_tag_before_default_cutoff, tt_tag_late], predicate=time@0 >= 631152000000000000, pruning_predicate=time_max@0 >= 631152000000000000
```

The use of `UnestExec` and CASE is suspicious)

Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: I think the issue is something related to `make_array` and nulls / no nulls learned about how to handle null values (namely in https://github.com/apache/arrow-datafusion/pull/6662). The `FilterExec: tagKey@0 IS NOT NULL` should take care of the nulls, but clearly it isn't working in this case. 

I will continue to dig

Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: I believe the issue is that `make_array` isn't treating nulls correctly. I'll work on a datafusion only reproducer. Good thing I was just talking about structured types with @pauldix 😆 

```
AAL make_array input:
[
    Array(
        StringArray
        [
          "tt_tag",
        ],
    ),
    Array(
        StringArray
        [
          "tt_tag_at_default_cutoff",
        ],
    ),
    Array(
        StringArray
        [
          null,
        ],
    ),
    Array(
        StringArray
        [
          "tt_tag_late",
        ],
    ),
]
AAL make_array result:
Ok(
    Array(
        ListArray
        [
          StringArray
        [
          "tt_tag",
          "tt_tag_at_default_cutoff",
          "",
          "tt_tag_late",
        ],
        ],
    ),
)
```

Comment (alamb) [reply influxdb_iox/tests/query_tests/cases/in/influxql_metadata.influxql.expected:1390]: Interestingly this looks like it has been fixed on the latest datafusion main. I will try and upgrade some more

**Pull request closed without merging**